### DISCUSSION TITLE
Windows Server 2025 - wmic Error

### DISCUSSION THREAD

**Discussion by mildblue** [rejetto/hfs#675] [Q&A] [answered]
![2024-07-09_165331](https://github.com/rejetto/hfs/assets/13603473/0b8ecd92-f3f4-4d55-8ff8-a6ccb05172a2)
![2024-07-09_165353](https://github.com/rejetto/hfs/assets/13603473/768e3ee7-55eb-4362-bcec-7e76ecfeee19)
![2024-07-09_210201](https://github.com/rejetto/hfs/assets/13603473/cdfa6366-c42d-4595-a102-2643c2228369)

An error occurs even though the 'wmic' service is working properly in Windows Server 2025. 
Isn't there a solution?

I downloaded "hfs-windows-x64-0.53.0-rc20" and tried it, but the same error occurs.

**Comment by rejetto** [comment]
i don't have windows 2025.
Maybe it doesn't have "chcp" command?
please copy and paste the command, the whole line, including chcp and wmic, on a command prompt, and see the result.

**Reply by mildblue** [reply]
> 저는 2025년 윈도우가 없습니다. "chcp" 명령어가 없는 것은 아닐까요? 명령 프롬프트에 chcp, wmic을 포함한 전체 행인 명령을 복사하여 붙여넣고 결과를 확인하십시오.

![2024-07-09_210041](https://github.com/rejetto/hfs/assets/13603473/64faf436-22c8-49af-8556-a6dcbc2b8ca9)


It's the same when you run it in administrator mode.

[Windows Server 2025 RTM.zip](https://github.com/user-attachments/files/16142542/Windows.Server.2025.RTM.zip)

**Comment by rejetto** [comment]
the error 44210 seems to be specific of wmic, i guess chcp has nothing to do with it.
try running as administrator, to see if that solves.

**Comment by rejetto** [comment, marked as answer]
hey, thanks but i cannot install Windows 2025 to make the test :)
thanks for the "administrator" test.
when i said the whole command, i meant "wmic service get ProcessId", not just "wmic". Let me know if you can try this, i'm interested in the result.
Also, i have a possible fix for your problem, it's included in this version:
https://github.com/user-attachments/files/16142024/hfs-windows-x64-0.53.0-rc21.1.zip
please let me know if it works.
it's just a workaround, in case the command fails, because it's not a critical problem: that command is necessary only for HFS to know if it is running as a service, because as a service it shouldn't attempt to auto-update. I repeat, it's not critical.

**Reply by mildblue** [reply]
![2024-07-09_204826](https://github.com/rejetto/hfs/assets/13603473/922b11c7-9d65-4477-bc49-d62abb7b1c51)


Thank you, I run it with this version and it works fine.

**Comment by rejetto** [comment]
thank you for your feedback!
that output is quite weird, something's wrong with it, as any system has services running, yours too, but it refuses to provide the list
<img width="271" alt="image" src="https://github.com/rejetto/hfs/assets/1367199/d9d29e1e-e496-4539-a668-bb9fdde0a12a">

but it's not even an empty list, because it doesn't include the first line, the one with "ProcessId", so i guess the command just failed, as it does when called by HFS.
I guess we'll live with that for now, but i have more attempts to do in future versions.

**Reply by mildblue** [reply]
Microsoft has recently made wmic.exe obsolete in Windows versions. Instead, we recommend using PowerShell. Therefore, it is likely that wmic.exe will be completely removed in future versions.
Here are some PowerShell commands you can use instead of wmic.exe:

Service Information Inquiry:

powershellCopyGet-CimInstance -ClassName Win32_Service | Select-Object Name, ProcessId

Inquire process information:

powershellCopyGet-Process | Select-Object Name, Id

Look up disk information:

powershellCopyGet-CimInstance -ClassName Win32_LogicalDisk

System Information Inquiry:

powershellCopyGet-CimInstance -ClassName Win32_ComputerSystem
These PowerShell commands can replace most of the features that wmic.exe provided.

^^;

**Reply by mildblue** [reply]
const { exec } = require('child_process');

const powershellCommand = 'Get-Process | Select-Object Name, Id | ConvertTo-Json';

exec(`powershell -Command "${powershellCommand}"`, (error, stdout, stderr) => {
  if (error) {
    console.error(`Error executing PowerShell command: ${error}`);
    return;
  }
  if (stderr) {
    console.error(`PowerShell command stderr: ${stderr}`);
    return;
  }

  try {
    const processes = JSON.parse(stdout);
    processes.forEach(process => {
      console.log(`Name: ${process.Name}, ID: ${process.Id}`);
    });
  } catch (parseError) {
    console.error(`Error parsing PowerShell output: ${parseError}`);
  }
});
